### DISCUSSION TITLE
Using sftp for

### DISCUSSION THREAD

**Discussion by ralmond** [quarto-dev/quarto-cli#3810] [Q&A]
I'm using the Quarto website feature to put together a simple website (computations are all frozen).  

To publish, I simply need to use `sftp` to recursively copy the `_site` directory to the right directory on the server.

I've got the ssh-key exchange set up, so I really only need to configure the name of the server and the directory on the server in which to copy the site.

Is there an easy way to do that with `quarto publish`?

**Comment by mcanouil** [comment]
Use `quarto render` to generate your site, then use any tool you want to upload the content of `_site`.

**Comment by ralmond** [comment]
Yes, but it would be nice if I could either, (a) set a couple of parameters and have `quarto publish` do the work for me or (b) write a shell script and have `quarto publish` call it.

**Reply by astrowonk** [reply]
I agree, but since that doesn't exist, I'd suggest an alias like `quartopublish` (or your name of choice) in bash/zsh that uses `rsync` over scp/sftp. Something like this:

```
rsync -avhz ./_site/ mywebserver.com:/the/path/to/my/webserver/dir/
```

would do the job with a similarly short amount of typing.

**Comment by mcanouil** [comment]
Quarto can't possibly be as general as you wish because it will in the end make Quarto a SFTP client which is not supposed to be.

`quarto publish` is only to make things easy for the more general hosts.

For everything else and as I said, use a post render script and `quarto render` this will give you all the flexibility you want and more (this is what `quarto publish` is kinda doing internally).

For short, what you need is `quarto render` not `quarto publish`.

**Reply by ralmond** [reply]
We may be quibbling over the meaning of "render" and "publish".

My typical workflow is as follows:
1) Make edits to the web site content.
2) `quarto render` and check that the results meet the requirements (if not, go back to 1)
3) `quarto publish`, i.e., copy rendered html and other files to the server for deployment.

I would be happy if I could attach the custom script to `quarto publish`.  `quarto render` is too soon as I have not completed my final checks yet.

**Reply by mcanouil** [reply]
You do not seem to understand and seem obsessed with the `quarto publish` command ...

> there is a `quarto publish` command available for easy publishing to various popular services (GitHub, Netlify, RStudio Connect, etc.)

> If you are publishing to a destination not listed above, choose [Other Services](https://quarto.org/docs/publishing/other.html).

> Publishing is simply a matter of copying the output directory to a web server or web hosting service.

As I said and as stated in the documentation, it can't do what you want and most likely never will, see <https://quarto.org/docs/publishing/other.html>.


So instead of your step 3, you have only two options:
- use a post-render script that perform the file/directory copy to where you want to host your content, see <https://quarto.org/docs/projects/scripts.html#pre-and-post-render>, possibly in association with a "project profile" to only render by default with `quarto render` and to render and publish with `quarto render --profile publish` (where `publish` is your custom yaml with the post-render option set with your script to copy the files)..
- Use the same script as above, but as a standalone.

**Reply by cscheid** [reply]
@ralmond If you're concerned about running your scripts when you don't intend to, you can use profiles: https://quarto.org/docs/projects/profiles.html By creating a `default` and an `upload` profile which includes the sftp scripts, you can distinguish between `quarto render --profile default`  and `quarto render --profile upload`. 

If you prefer, you can also use environment variables (which will be passed to all scripts). Then use `UPLOAD=true quarto render`, and have your scripts do nothing when `UPLOAD` is not set.
